relays/ethereum/src/substrate_client.rs

```diff
@@ -224,42 +308,30 @@ async fn next_account_index(
 ) -> (Client, Result<node_primitives::Index, Error>) {
 	use sp_core::crypto::Ss58Codec;
 
-	let (client, index) = call_rpc_u64(
-		client,
-		"system_accountNextIndex",
-		Params::Array(vec![to_value(account.to_ss58check()).unwrap()]),
-	)
+	let account = bail_on_arg_error!(
+		to_value(account.to_ss58check()).map_err(|e| Error::RequestSerialization(e)),
+		client
+	);
+	let (client, index) = call_rpc(client, "system_accountNextIndex", Params::Array(vec![account]), |v| {
+		rpc_returns_value::<u64>(v)
+	})
 	.await;
 	(client, index.map(|index| index as _))
 }
 
 /// Calls RPC on Substrate node that returns Bytes.
-async fn call_rpc<T: Decode>(mut client: Client, method: &'static str, params: Params) -> (Client, Result<T, Error>) {
-	async fn do_call_rpc<T: Decode>(client: &mut Client, method: &'static str, params: Params) -> Result<T, Error> {
-		let request_id = client
-			.rpc_client
-			.start_request(method, params)
-			.await
-			.map_err(Error::StartRequestFailed)?;
-		// WARN: if there'll be need for executing >1 request at a time, we should avoid
-		// calling request_by_id
-		let response = client
-			.rpc_client
-			.request_by_id(request_id)
-			.ok_or(Error::RequestNotFound)?
-			.await
-			.map_err(Error::ResponseRetrievalFailed)?;
-		let encoded_response: Bytes = from_value(response).map_err(|_| Error::ResponseParseFailed)?;
-		Decode::decode(&mut &encoded_response.0[..]).map_err(|_| Error::ResponseParseFailed)
-	}
-
-	let result = do_call_rpc(&mut client, method, params).await;
-	(client, result)
-}
-
-/// Calls RPC on Substrate node that returns u64.
-async fn call_rpc_u64(mut client: Client, method: &'static str, params: Params) -> (Client, Result<u64, Error>) {
-	async fn do_call_rpc(client: &mut Client, method: &'static str, params: Params) -> Result<u64, Error> {
+async fn call_rpc<T>(
+	mut client: Client,
+	method: &'static str,
+	params: Params,
+	decode_value: impl Fn(Value) -> Result<T, Error>,
+) -> (Client, Result<T, Error>) {
+	async fn do_call_rpc<T>(
+		client: &mut Client,
```

Comment:
Why do we keep passing owned Client if a mutable reference is enough? It's because we then spawn the futures on an executor? Maybe best to just put these bare functions to a struct where `Client` is owned?